# ETL
This notebook loads in the dataset, enforces correct typecasting, adds a column containing all of the duplicates for each row, and saves the dataset as a pickle file.

In [1]:
import pandas as pd

In [2]:
import os
print("Current Working Directory: ", os.getcwd())

Current Working Directory:  /home/msa242/CMPT413-Final-Project/ETL


### Read in Data

In [3]:
dataset_name = 'mozilla_firefox'
df = pd.read_csv('../data/' + dataset_name + '.csv')
df.head()

Issue_id Priority    Component  Duplicated_issue  \
0     10954       P3  Preferences               NaN   
1     14871       --      General          269442.0   
2     19118       --  Preferences               NaN   
3     54746       P3  Preferences               NaN   
4     56892       P3      General               NaN   

                                               Title  \
0     Dialup properties needs to be exposed in prefs   
1                        [Find] Find whole word only   
2  Plug-In Manager (ui for choosing mimetype-plug...   
3  Language encodings in font prefs dialog not so...   
4           Synaptics touchpad scrolling not working   

                                         Description    Status  Resolution  \
0  The dialup properties of the profile should be...  RESOLVED     WONTFIX   
1  Please add Match Whole Word Only option to bro...  RESOLVED   DUPLICATE   
2  I would really like a plug-in manager for my b...  RESOLVED     WONTFIX   
3  Language encodings are listed in a seemingly r...  RESOLVED  WORKSFORME   
4  From Bugzilla Helper:; User-Agent: Mozilla/5.0...  RESOLVED  WORKSFORME   

       Version               Created_time              Resolved_time  
0        Trunk  1999-07-30 15:55:51 -0700  2008-05-14 11:44:15 -0700  
1        Trunk  1999-09-24 14:49:34 -0700  2011-10-05 16:35:31 -0700  
2        Trunk  1999-11-17 14:58:26 -0800  2013-01-29 11:48:39 -0800  
3        Trunk  2000-09-29 14:12:11 -0700  2013-02-27 15:47:29 -0800  
4  unspecified  2000-10-16 14:48:15 -0700  2009-10-14 11:38:29 -0700

In [4]:
df.shape

(115814, 11)

### Typecast

In [5]:
def typecast_df(df):
    df['Content'] = df['Title'] + ' ' + df['Description']
    df['Duplicated_issue'] = pd.to_numeric(df['Duplicated_issue'], errors='coerce').astype('Int64')
    df["Issue_id"] = df["Issue_id"].astype('Int64')
    return df
df = typecast_df(df)

In [6]:
df.head()

Issue_id Priority    Component  Duplicated_issue  \
0     10954       P3  Preferences              <NA>   
1     14871       --      General            269442   
2     19118       --  Preferences              <NA>   
3     54746       P3  Preferences              <NA>   
4     56892       P3      General              <NA>   

                                               Title  \
0     Dialup properties needs to be exposed in prefs   
1                        [Find] Find whole word only   
2  Plug-In Manager (ui for choosing mimetype-plug...   
3  Language encodings in font prefs dialog not so...   
4           Synaptics touchpad scrolling not working   

                                         Description    Status  Resolution  \
0  The dialup properties of the profile should be...  RESOLVED     WONTFIX   
1  Please add Match Whole Word Only option to bro...  RESOLVED   DUPLICATE   
2  I would really like a plug-in manager for my b...  RESOLVED     WONTFIX   
3  Language encodings are listed in a seemingly r...  RESOLVED  WORKSFORME   
4  From Bugzilla Helper:; User-Agent: Mozilla/5.0...  RESOLVED  WORKSFORME   

       Version               Created_time              Resolved_time  \
0        Trunk  1999-07-30 15:55:51 -0700  2008-05-14 11:44:15 -0700   
1        Trunk  1999-09-24 14:49:34 -0700  2011-10-05 16:35:31 -0700   
2        Trunk  1999-11-17 14:58:26 -0800  2013-01-29 11:48:39 -0800   
3        Trunk  2000-09-29 14:12:11 -0700  2013-02-27 15:47:29 -0800   
4  unspecified  2000-10-16 14:48:15 -0700  2009-10-14 11:38:29 -0700   

                                             Content  
0  Dialup properties needs to be exposed in prefs...  
1  [Find] Find whole word only Please add Match W...  
2  Plug-In Manager (ui for choosing mimetype-plug...  
3  Language encodings in font prefs dialog not so...  
4  Synaptics touchpad scrolling not working From ...

### Find Duplicated_issue ids that do not have corresponding entries

In [7]:
missing_issues = df[~df['Duplicated_issue'].isna() & ~df['Duplicated_issue'].isin(df['Issue_id'])]
missing_issues

Issue_id Priority  Component  Duplicated_issue  \
1          14871       --    General            269442   
64        167422       --    General            162593   
92        169727       --    General            126730   
101       170029       --      Menus             25894   
106       170216       --    General            280190   
...          ...      ...        ...               ...   
115785    953147       --   Security            616853   
115790    953223       --      Menus            787875   
115795    953332       --  Untriaged            248955   
115801    953436       --  Untriaged             50673   
115808    955868       --   Security            636374   

                                                    Title  \
1                             [Find] Find whole word only   
64      view source tries to download type application...   
92      Restored window position and size lost after a...   
101     Respect Windows Access Keys display pref (defa...   
106                        Pretty up dns error page a bit   
...                                                   ...   
115785  Ransomware locks Firefox tab; uses onbeforeunl...   
115790  right clicking on an image should offer Google...   
115795                   Per-tab search engine search bar   
115801                partial copying of hyperlinked text   
115808  Block ransomware: Limit rate of confirm to lea...   

                                              Description    Status  \
1       Please add Match Whole Word Only option to bro...  RESOLVED   
64      view source treis to download application x-vi...  VERIFIED   
92      Tested on: Mozilla/5.0 (X11; U; Linux i686; en...  RESOLVED   
101           only shows the access keys when you hit alt  RESOLVED   
106                                                        VERIFIED   
...                                                   ...       ...   
115785  Created attachment 8351469; HTML from the site...  RESOLVED   
115790  Created attachment 8351530; f.png; ; User Agen...  RESOLVED   
115795  User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64...  RESOLVED   
115801  User Agent: Mozilla/5.0 (X11; Linux i686; rv:2...  RESOLVED   
115808  Created attachment 8355009; source of the rans...  RESOLVED   

       Resolution      Version               Created_time  \
1       DUPLICATE        Trunk  1999-09-24 14:49:34 -0700   
64      DUPLICATE  unspecified  2002-09-08 18:06:09 -0700   
92      DUPLICATE        Trunk  2002-09-19 13:01:39 -0700   
101     DUPLICATE  unspecified  2002-09-20 23:20:38 -0700   
106     DUPLICATE  unspecified  2002-09-22 14:24:45 -0700   
...           ...          ...                        ...   
115785  DUPLICATE        Trunk  2013-12-25 21:14:32 -0800   
115790  DUPLICATE    28 Branch  2013-12-26 16:58:37 -0800   
115795  DUPLICATE    26 Branch  2013-12-28 03:24:38 -0800   
115801  DUPLICATE    28 Branch  2013-12-29 16:34:33 -0800   
115808  DUPLICATE    26 Branch  2013-12-31 08:43:52 -0800   

                    Resolved_time  \
1       2011-10-05 16:35:31 -0700   
64      2002-09-19 13:24:25 -0700   
92      2008-04-06 11:55:18 -0700   
101     2009-03-30 13:54:45 -0700   
106     2012-06-04 19:34:48 -0700   
...                           ...   
115785  2013-12-26 17:48:18 -0800   
115790  2013-12-27 10:49:27 -0800   
115795  2013-12-28 06:08:41 -0800   
115801  2013-12-30 00:21:36 -0800   
115808  2014-01-01 08:36:52 -0800   

                                                  Content  
1       [Find] Find whole word only Please add Match W...  
64      view source tries to download type application...  
92      Restored window position and size lost after a...  
101     Respect Windows Access Keys display pref (defa...  
106                      Pretty up dns error page a bit    
...                                                   ...  
115785  Ransomware locks Firefox tab; uses onbeforeunl...  
115790  right clicking on an image should offer Google...  
115795 

### Create Duplicated_issue column containing an array of Duplicated_issue ids

In [8]:
duplicates_dict = df.groupby('Issue_id')['Duplicated_issue'].apply(set).to_dict()

def get_all_duplicates(issue_id):
    duplicates = set()
    issues_to_check = [issue_id]

    while issues_to_check:
        current_issue = issues_to_check.pop()
        current_duplicates = duplicates_dict.get(current_issue, set())

        new_duplicates = current_duplicates - duplicates
        duplicates.update(new_duplicates)
        issues_to_check.extend(new_duplicates)

    return list(duplicates) if duplicates else []

df['Duplicated_issues'] = df['Issue_id'].apply(get_all_duplicates)

In [9]:
def clean_convert_array(arr):
    cleaned = [int(x) for x in arr if not pd.isnull(x)]
    return cleaned

df['Duplicated_issues'] = df['Duplicated_issues'].apply(clean_convert_array)
df = df.sort_values(by='Duplicated_issues', key=lambda x: x.str.len(), ascending=False)

In [10]:
df.head()

Issue_id Priority                   Component  Duplicated_issue  \
33591     335190       --                     General            335186   
21703     294616       --                       Menus            236336   
104453    787029       --                   Untriaged            787022   
1900      203901       --  Toolbars and Customization            203899   
1898      203898       --  Toolbars and Customization            203899   

                                                    Title  \
33591                                Ctrl+C does not copy   
21703                            Open URL in context menu   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

                                              Description    Status  \
33591   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
21703   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
104453  User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64...  RESOLVED   
1900    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   
1898    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   

       Resolution      Version               Created_time  \
33591   DUPLICATE  unspecified  2006-04-23 18:45:42 -0700   
21703   DUPLICATE  unspecified  2005-05-18 02:59:14 -0700   
104453  DUPLICATE    14 Branch  2012-08-30 06:12:03 -0700   
1900    DUPLICATE  unspecified  2003-04-30 03:21:59 -0700   
1898    DUPLICATE  unspecified  2003-04-30 03:10:53 -0700   

                    Resolved_time  \
33591   2006-04-23 19:50:22 -0700   
21703   2006-01-19 00:20:44 -0800   
104453  2012-08-30 06:30:58 -0700   
1900    2006-11-13 07:16:57 -0800   
1898    2006-11-13 07:23:00 -0800   

                                                  Content  \
33591   Ctrl+C does not copy User-Agent:       Mozilla...   
21703   Open URL in context menu User-Agent:       Moz...   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

               Duplicated_issues  
33591   [335186, 334500, 334862]  
21703   [236336, 227922, 454518]  
104453  [651803, 787021, 787022]  
1900    [197227, 203899, 210910]  
1898    [197227, 203899, 210910]

### Count the number of issues with duplicates before removing nonexistent id's

In [11]:
num_issues_with_duplicates = df['Duplicated_issues'].apply(lambda x: len(x) > 0).sum()
num_issues_with_duplicates

35814

### Remove the nonexistent id's from Duplicate_issues arrays

In [12]:
issue_ids = set(df['Issue_id'])

def remove_nonexistent_issue_ids(duplicated_issues):
    return [issue_id for issue_id in duplicated_issues if issue_id in issue_ids]

df['Duplicated_issues'] = df['Duplicated_issues'].apply(remove_nonexistent_issue_ids)
df.head()

Issue_id Priority                   Component  Duplicated_issue  \
33591     335190       --                     General            335186   
21703     294616       --                       Menus            236336   
104453    787029       --                   Untriaged            787022   
1900      203901       --  Toolbars and Customization            203899   
1898      203898       --  Toolbars and Customization            203899   

                                                    Title  \
33591                                Ctrl+C does not copy   
21703                            Open URL in context menu   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

                                              Description    Status  \
33591   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
21703   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
104453  User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64...  RESOLVED   
1900    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   
1898    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   

       Resolution      Version               Created_time  \
33591   DUPLICATE  unspecified  2006-04-23 18:45:42 -0700   
21703   DUPLICATE  unspecified  2005-05-18 02:59:14 -0700   
104453  DUPLICATE    14 Branch  2012-08-30 06:12:03 -0700   
1900    DUPLICATE  unspecified  2003-04-30 03:21:59 -0700   
1898    DUPLICATE  unspecified  2003-04-30 03:10:53 -0700   

                    Resolved_time  \
33591   2006-04-23 19:50:22 -0700   
21703   2006-01-19 00:20:44 -0800   
104453  2012-08-30 06:30:58 -0700   
1900    2006-11-13 07:16:57 -0800   
1898    2006-11-13 07:23:00 -0800   

                                                  Content  \
33591   Ctrl+C does not copy User-Agent:       Mozilla...   
21703   Open URL in context menu User-Agent:       Moz...   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

               Duplicated_issues  
33591           [335186, 334862]  
21703   [236336, 227922, 454518]  
104453  [651803, 787021, 787022]  
1900    [197227, 203899, 210910]  
1898    [197227, 203899, 210910]

In [13]:
num_issues_with_duplicates = df['Duplicated_issues'].apply(lambda x: len(x) > 0).sum()
num_issues_with_duplicates

15742

After removing the non-existent issue ids, we reduce the number of issues with duplicates from $35814$ to $15742$, a reduction by $56\%$

### Clean up
- Remove the "Duplicated_issue" column
- Set "Resolution" with no duplicates and Resolution==DUPLICATE to Resolution==NDUPLICATE
- Add "Duplicates_count" column

In [14]:
df = df.drop(columns=['Duplicated_issue'])
df.loc[(df['Duplicated_issues'].str.len() == 0) & (df['Resolution'] == 'DUPLICATE'), 'Resolution'] = 'NDUPLICATE'
df['Duplicates_count'] = df['Duplicated_issues'].apply(len)
df.head()

Issue_id Priority                   Component  \
33591     335190       --                     General   
21703     294616       --                       Menus   
104453    787029       --                   Untriaged   
1900      203901       --  Toolbars and Customization   
1898      203898       --  Toolbars and Customization   

                                                    Title  \
33591                                Ctrl+C does not copy   
21703                            Open URL in context menu   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

                                              Description    Status  \
33591   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
21703   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
104453  User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64...  RESOLVED   
1900    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   
1898    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   

       Resolution      Version               Created_time  \
33591   DUPLICATE  unspecified  2006-04-23 18:45:42 -0700   
21703   DUPLICATE  unspecified  2005-05-18 02:59:14 -0700   
104453  DUPLICATE    14 Branch  2012-08-30 06:12:03 -0700   
1900    DUPLICATE  unspecified  2003-04-30 03:21:59 -0700   
1898    DUPLICATE  unspecified  2003-04-30 03:10:53 -0700   

                    Resolved_time  \
33591   2006-04-23 19:50:22 -0700   
21703   2006-01-19 00:20:44 -0800   
104453  2012-08-30 06:30:58 -0700   
1900    2006-11-13 07:16:57 -0800   
1898    2006-11-13 07:23:00 -0800   

                                                  Content  \
33591   Ctrl+C does not copy User-Agent:       Mozilla...   
21703   Open URL in context menu User-Agent:       Moz...   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

               Duplicated_issues  Duplicates_count  
33591           [335186, 334862]                 2  
21703   [236336, 227922, 454518]                 3  
104453  [651803, 787021, 787022]                 3  
1900    [197227, 203899, 210910]                 3  
1898    [197227, 203899, 210910]                 3

### Save as Pickle File

In [15]:
df.to_pickle('../data/' + dataset_name + '.pkl')

### Read the pickle file into a dataframe

In [16]:
df = pd.read_pickle('../data/' + dataset_name + '.pkl')
df.head()

Issue_id Priority                   Component  \
33591     335190       --                     General   
21703     294616       --                       Menus   
104453    787029       --                   Untriaged   
1900      203901       --  Toolbars and Customization   
1898      203898       --  Toolbars and Customization   

                                                    Title  \
33591                                Ctrl+C does not copy   
21703                            Open URL in context menu   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

                                              Description    Status  \
33591   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
21703   User-Agent:       Mozilla/5.0 (Windows; U; Win...  RESOLVED   
104453  User Agent: Mozilla/5.0 (Windows NT 6.1; WOW64...  RESOLVED   
1900    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   
1898    User-Agent:       Mozilla/5.0 (Windows; U; Win...  VERIFIED   

       Resolution      Version               Created_time  \
33591   DUPLICATE  unspecified  2006-04-23 18:45:42 -0700   
21703   DUPLICATE  unspecified  2005-05-18 02:59:14 -0700   
104453  DUPLICATE    14 Branch  2012-08-30 06:12:03 -0700   
1900    DUPLICATE  unspecified  2003-04-30 03:21:59 -0700   
1898    DUPLICATE  unspecified  2003-04-30 03:10:53 -0700   

                    Resolved_time  \
33591   2006-04-23 19:50:22 -0700   
21703   2006-01-19 00:20:44 -0800   
104453  2012-08-30 06:30:58 -0700   
1900    2006-11-13 07:16:57 -0800   
1898    2006-11-13 07:23:00 -0800   

                                                  Content  \
33591   Ctrl+C does not copy User-Agent:       Mozilla...   
21703   Open URL in context menu User-Agent:       Moz...   
104453  movement was canceled in web page   Navigation...   
1900    Folders in Bookmarks Toolbar stay Expanded if ...   
1898    Folders in Bookmarks Toolbar stay Expanded if ...   

               Duplicated_issues  Duplicates_count  
33591           [335186, 334862]                 2  
21703   [236336, 227922, 454518]                 3  
104453  [651803, 787021, 787022]                 3  
1900    [197227, 203899, 210910]                 3  
1898    [197227, 203899, 210910]                 3